components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -709,26 +687,104 @@ private boolean isDomainNameEmbeddedInAttributeValue(String filterOperation, Str
      * @return User names of the filtered users
      * @throws CharonException Error while filtering
      */
-    private String[] getFilteredUsers(Node node, int offset, int limit, String sortBy, String sortOrder,
-            String domainName) throws CharonException {
+    private String[] filterUsers(Node node, int offset, int limit, String sortBy, String sortOrder, String domainName)
+            throws CharonException {
 
+        // Filter users when the domain is specified in the request.
         if (StringUtils.isNotEmpty(domainName)) {
-            return getFilteredUsernames(node, offset, limit, sortBy, sortOrder, domainName);
+            return getFilteredUsernames(createConditionForSingleAttributeFilter(domainName, node), offset, limit,
+                    sortBy, sortOrder, domainName);
         } else {
+            // Filter users when the domain is not set in the request. Then filter through multiple domains.
             String[] userStoreDomainNames = getDomainNames();
             ArrayList<String> filteredUsernames = new ArrayList<>();
             for (String userStoreDomainName : userStoreDomainNames) {
-                filteredUsernames.addAll(Arrays
-                        .asList(getFilteredUsernames(node, offset, limit, sortBy, sortOrder, userStoreDomainName)));
+
+                // Create filter condition for each domain.
+                Condition condition = createConditionForSingleAttributeFilter(domainName, node);
+
+                // Filter users for given condition and domain.
+                String[] userNames = getFilteredUsernames(condition, offset, limit, sortBy, sortOrder,
+                        userStoreDomainName);
+                if (userNames == null) {
+                    userNames = new String[0];
+                }
+
+                // Calculating new offset and limit parameters.
+                int numberOfFilteredUsers = userNames.length;
+                if (numberOfFilteredUsers <= 0 && offset > 1) {
+                    offset = calculateOffset(condition, offset, sortBy, sortOrder, domainName);
+                } else {
+                    // Returned user names size > 0 implies there are users in that domain which is larger than the offset.
```

Comment:
this line exceeds the max allowed character length (120)

Reply:
Fixed: f999948935ff2081b9362c4cbe0d6e8f5ee238db